File: 27_leak_canary/src/main/java/com/huoergai/leak_canary/MainActivity.java
package com.huoergai.leak_canary;

import android.os.Bundle;
import android.util.Log;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Util.requestPermission(this);
        setContentView(R.layout.activity_main);
    }

    @SuppressWarnings({"AccessStaticViaInstance"})
    public void onClick(View view) {
        switch (view.getId()) {
            // 静态变量
            case R.id.static_variable:
                Util.leakViews.add(view);
                break;
            // 局部变量
            case R.id.stack:
                MethodStack.startWithTarget(view);
                break;
            // 线程对象
            case R.id.thread:
                LeakThread leakThread = new LeakThread();
                leakThread.leakViews.add(view);
                leakThread.start();
                break;
            // application
            case R.id.application:
                ((App) getApplication()).leakViews.add(view);
                break;
            // dump
            case R.id.dump:
                Util.dump(this);
                break;
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Log.d("MainActivity", "onDestory.....");
        // Util.gc();
    }

    @Override
    protected void finalize() throws Throwable {
        super.finalize();
        Log.d("MainActivity", "finalize......");
    }
}

File: 11_text_geo_transfer/src/main/java/com/huoergai/text/MainActivity.kt
package com.huoergai.text

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity

/**
 * 文字测量和几何变换（text measurement & geometry transformation）
 *
 * 几何变换：translate rotate scale skew
 * 变换的时候为方便理清思路，使用逆序和操作视图而不是坐标系。
 */
class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
    }
}
